src/Storages/MergeTree/MergeTreeData.cpp

```diff
@@ -2153,10 +2148,33 @@ void MergeTreeData::dropAllData()
     if (!getStorageID().hasUUID())
         getContext()->dropCaches();
 
-    LOG_TRACE(log, "dropAllData: removing data from filesystem.");
 
     /// Removing of each data part before recursive removal of directory is to speed-up removal, because there will be less number of syscalls.
-    clearPartsFromFilesystem(all_parts);
+    NameSet part_names_failed;
+    try
+    {
+        LOG_TRACE(log, "dropAllData: removing data parts (count {}) from filesystem.", all_parts.size());
+        clearPartsFromFilesystem(all_parts, true, &part_names_failed);
+
+        LOG_TRACE(log, "dropAllData: removing all data parts from memory.");
+        data_parts_indexes.clear();
+    }
+    catch (...)
+    {
+        /// Removing from memory only successfully removed parts from disk
+        /// Parts removal process can be important and on the next try it's better to try to remove
+        /// them instead of remove recursive call.
+        LOG_WARNING(log, "dropAllData: got exception removing parts from disk, removing successfully removed parts from memory.");
+        for (const auto & part : all_parts)
+        {
+            if (!part_names_failed.contains(part->name))
+                data_parts_indexes.erase(part->info);
+        }
+
+        throw;
```

Comment:
FYI it may cause some issues with Ordinary database (but I don't care)

Reply:
Do we still need `removeRecursive` below?

Reply:
I don't know, only as safe-guard against some garbage like /detached?